### PR TITLE
CMM-1288: Add endpoints for Subscribers tab

### PR DESCRIPTION
**Pull request by adalpari**
## Description

Add three new WP.com API endpoints to support the Subscribers tab feature:
1. **Stats emails summary** — `/rest/v1.1/sites/{site_id}/stats/emails/summary`
2. **Stats subscribers** — `/rest/v1.1/sites/{site_id}/stats/subscribers`
3. **Subscribers by user type** — `/wpcom/v2/sites/{site_id}/subscribers_by_user_type`

### Stats emails summary
- Add `StatsEmailsSummaryParams` with `period`, `quantity`, `sort_field`, and `sort_order` parameters
- Add `StatsEmailsSummaryResponse` and `StatsEmailsSummaryPost` response types
- Add endpoint definition using `WpDerivedRequest` macro with `WpComNamespace::RestV1_1`
- Wire up the endpoint in `WpComApiRequestBuilder` and `WpComApiClient`
- Add 10 unit tests and 3 JSON test fixtures
- Add e2e integration test

### Stats subscribers
- Add `StatsSubscribersParams` with `unit`, `quantity`, and `stat_fields` (comma-joined) parameters
- Add `StatsSubscribersResponse` reusing `StatsVisitsDataValue` for the fields+data array pattern
- Add helper methods `subscribers_data()` and `subscribers_paid_data()` with `#[uniffi::export]`
- Add endpoint definition with `WpComNamespace::RestV1_1`
- Wire up the endpoint in `WpComApiRequestBuilder` and `WpComApiClient`
- Add 13 unit tests and 5 JSON test fixtures (day/week/month/year/empty)
- Add e2e integration test

### Subscribers by user type
- Add `SubscribersByUserTypeParams` with `user_type`, `per_page`, `page`, and `sort` parameters
- Add `SubscribersByUserTypeUserType` and `SubscribersByUserTypeSortField` enums
- Reuse existing `ListSubscribersResponse` type
- Add `ListSubscribersByUserType` variant to `SubscribersRequest` with `WpComNamespace::V2`
- Add 4 unit tests and 2 JSON test fixtures

## Test plan

- [x] `cargo build` succeeds
- [x] `cargo test --lib -p wp_api` — all new unit tests pass
- [x] `cargo fmt` clean
- [ ] Run e2e tests with WP.com credentials

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: wp_api/src/wp_com/stats_emails_summary.rs [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+)]
+#[serde(rename_all = "snake_case")]
+#[strum(serialize_all = "snake_case")]
+pub enum StatsEmailsSummaryPeriod {
+    Day,
+    Week,
+    Month,
+    Year,
+    #[default]
+    Alltime,
```
Comment (jkmassel) [inline wp_api/src/wp_com/stats_emails_summary.rs]: optional nit:


```suggestion
    AllTime,
```

File: wp_api/src/wp_com/subscribers.rs [hunk 2]
```diff
@@ -5,18 +5,26 @@ use crate::{
...
 use std::{collections::HashMap, ops::Not};
 use wp_serde_helper::{deserialize_false_or_string, deserialize_u64_or_string};
 
+fn deserialize_date_subscribed<'de, D>(deserializer: D) -> Result<Option<WpGmtDateTime>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    Ok(WpGmtDateTime::deserialize(deserializer).ok())
+}
+
```
Comment (jkmassel) [inline wp_api/src/wp_com/subscribers.rs]: I don't think we want a custom deserializer here – can we use put a `wp_utc_date_format_optional` deserializer alongside `wp_utc_date_format`?

File: wp_api/src/wp_com/subscribers.rs [hunk 3]
```diff
@@ -5,18 +5,26 @@ use crate::{
...
 #[derive(Debug, Serialize, Deserialize, uniffi::Record)]
 pub struct Subscriber {
     pub user_id: UserId,
     pub subscription_id: SubscriptionId,
     pub display_name: String,
     pub email_address: String,
     pub is_email_subscriber: bool,
-    pub date_subscribed: WpGmtDateTime,
+    #[serde(deserialize_with = "deserialize_date_subscribed")]
+    pub date_subscribed: Option<WpGmtDateTime>,
```
Comment (jkmassel) [inline wp_api/src/wp_com/subscribers.rs]: Building on the comment above, I think you need `#[serde(default)]` to coerce the value to `None` if the value is missing entirely.

File: wp_api/src/wp_com/stats_subscribers.rs [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+    let field_index = match response.fields.iter().position(|field| field == handle) {
+        Some(index) => index,
+        None => return vec![],
+    };
+
+    response
+        .data
+        .iter()
+        .filter_map(|row| {
+            if let Some(period) = row[0].as_string()
```
Comment (jkmassel) [inline wp_api/src/wp_com/stats_subscribers.rs]: The bot flagged this:

> Two issues here:
>
> 1. `row[0]` hard-codes the assumption that the period field is always at index 0. The data field is looked up dynamically via `position()`, but the period field isn't — if the API ever returns `fields` in a different order, all rows get silently filtered out.
> 2. Both `row[0]` and `row[field_index]` use direct indexing, which panics if a data row is shorter than expected.

> Suggested fix:

```rust
let period_index = match response.fields.iter().position(|f| f == "period") {
    Some(i) => i,
    None => return vec![],
};

let field_index = match response.fields.iter().position(|field| field == handle) {
    Some(index) => index,
    None => return vec![],
};

response
    .data
    .iter()
    .filter_map(|row| {
        if let Some(period) = row.get(period_index).and_then(|v| v.as_string())
            && let Some(value) = row.get(field_index).and_then(|v| v.as_number())
        {
            return Some((period.clone(), value));
        }
        None
    })
    .collect()
```

The same issue exists in `stats_visits.rs` and we should address it there too (can be a separate PR).

File: wp_api/src/wp_com/subscribers.rs (hunk 3 above)
Comment (jkmassel) [reply wp_api/src/wp_com/subscribers.rs]: I assume the new endpoint has this potentially optional while the old one didn't, which is the reason for the optionality change?

File: wp_api/src/wp_com/endpoint/subscribers_endpoint.rs [hunk 5]
```diff
@@ -28,6 +28,8 @@ enum SubscribersRequest {
     AddSubscribers,
     #[get(url = "/sites/<wp_com_site_id>/subscribers/stats", output = SubscriberStatsResponse)]
     GetSubscriberStats,
+    #[get(url = "/sites/<wp_com_site_id>/subscribers_by_user_type", params = &SubscribersByUserTypeParams, output = ListSubscribersResponse)]
```
Comment (jkmassel) [inline wp_api/src/wp_com/endpoint/subscribers_endpoint.rs:31]: Do we have an e2e test for this?

File: wp_api/src/wp_com/stats_emails_summary.rs [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+    Hash,
+    Serialize,
+    Deserialize,
+    uniffi::Enum,
+    strum_macros::EnumString,
+    strum_macros::Display,
+)]
+#[serde(rename_all = "snake_case")]
+#[strum(serialize_all = "snake_case")]
+pub enum StatsEmailsSummarySortOrder {
```
Comment (jkmassel) [inline wp_api/src/wp_com/stats_emails_summary.rs]: WDYT about using `WpApiParamOrder` instead?

File: wp_api/src/wp_com/subscribers.rs [hunk 7]
```diff
@@ -198,6 +206,95 @@ pub struct ListSubscribersResponse {
...
+    Ord,
+    Hash,
+    Serialize,
+    Deserialize,
+    uniffi::Enum,
+    strum_macros::EnumString,
+    strum_macros::Display,
+)]
+#[strum(serialize_all = "snake_case")]
+pub enum SubscribersByUserTypeSortField {
```
Comment (jkmassel) [inline wp_api/src/wp_com/subscribers.rs:253]: Nit: This is missing `#[serde(rename_all = "snake_case")]` — the convention elsewhere in the crate is to have both `#[serde(rename_all = "snake_case")]` and `#[strum(serialize_all = "snake_case")]` together. The existing `ListSubscribersSortField` and `ListSubscribersOrder` above have the same omission, but the rest of the `wp_com` stats enums all include both.

File: wp_api/src/wp_com/stats_subscribers.rs [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+                .map(|f| f.to_string())
+                .collect::<Vec<_>>()
+                .join(",");
+            query_pairs_mut.append_query_value_pair("stat_fields", &stat_fields_str);
+        }
+    }
+}
+
+/// Response from the stats subscribers endpoint.
+#[derive(Debug, Hash, Serialize, Deserialize, uniffi::Record)]
```
Comment (jkmassel) [inline wp_api/src/wp_com/stats_subscribers.rs]: What does this need `Hash` for?

File: wp_api/src/wp_com/subscribers.rs (hunk 3 above)
Comment (jkmassel) [reply wp_api/src/wp_com/subscribers.rs]: If this can be optional, can we have a JSON test fixture that demonstrates this?

**Review by jkmassel — CHANGES_REQUESTED** [review]
Caught a few things to take a look over (mostly suggestions by the bot)

File: wp_api/src/wp_com/subscribers.rs (hunk 3 above)
Comment (adalpari) [reply wp_api/src/wp_com/subscribers.rs]: > I assume the new endpoint has this potentially optional while the old one didn't, which is the reason for the optionality change?

There was a case where I got inconsistent `date_subscribed` (see: https://github.com/Automattic/wordpress-rs/pull/1200/changes/b17d6e02dde5ea55773d527c1cb6ad2f889fc90e), but you are right, I didn't add tests for it. Let me investigate it

File: wp_api/src/wp_com/stats_subscribers.rs (hunk 4 above)
Comment (adalpari) [reply wp_api/src/wp_com/stats_subscribers.rs]: Good catch!!
Will open another PR for `stats_visits.rs` as well

File: wp_api/src/wp_com/stats_emails_summary.rs (hunk 6 above)
Comment (adalpari) [reply wp_api/src/wp_com/stats_emails_summary.rs]: Good call!!

File: wp_api/src/wp_com/stats_subscribers.rs (hunk 8 above)
Comment (adalpari) [reply wp_api/src/wp_com/stats_subscribers.rs]: It seems this was added because of following other endpoints ´derive´.  I missed it.

I am removing it and also investigating from where it was copied and if it's necessary at all in the other endpoints. Will include it in the PR related to [this comment](https://github.com/Automattic/wordpress-rs/pull/1200#discussion_r2885745314).

File: wp_api/src/wp_com/subscribers.rs (hunk 3 above)
Comment (adalpari) [reply wp_api/src/wp_com/subscribers.rs]: I'm not able to reproduce the error anymore, but tracking the interaction this is what I got:

> ResponseParsingError(reason=Invalid date format: -001-11-30T00:00:00+00:00 at line 1 column 41532, response={"

Just added a [test](https://github.com/Automattic/wordpress-rs/pull/1200/commits/3a23a7389b3931e6241986a168b73a95d99151e0) to cover the case.

**Comment by adalpari** [comment]
@jkmassel can you take another look?

File: wp_api/tests/wpcom/subscribers/subscribers-by-user-type-with-invalid-date.json [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+            "subscription_status": null,
+            "avatar": "https://0.gravatar.com/avatar/example1",
+            "display_name": "Nik",
+            "url": "https://nikhilc.dev"
+        },
+        {
+            "user_id": 3251794,
+            "subscription_id": 10660,
+            "email_address": "olduser@example.com",
+            "date_subscribed": "-001-11-30T00:00:00+00:00",
```
Comment (adalpari) [inline wp_api/tests/wpcom/subscribers/subscribers-by-user-type-with-invalid-date.json:22]: Note: This format comes from this real error I got while using the app:

`ResponseParsingError(reason=Invalid date format: -001-11-30T00:00:00+00:00 at line 1 column 41532, response={"total":88,"pages":1,"page":1,"per_page":100,"subscribers":[{"user_id":33840434,"subscription_id":792200219,"email_address":"email@nikhilchavan.com","date_subscribed":"2024-12-26T10:08:55+00:00","is_email_subscriber":false,"subscription_status":null,"avatar":"https:\/\/0.gravatar.com\/avatar\/fb265c001fcd9384d6dc59d2c47f43196c3a5835e298bcd0ba18b5ae0d82bb96?s=128&amp;d=https%3A%2F%2F0.gravatar.com%2Favatar%2Fad516503a11cd5ca435acc9bb6523536%3Fs%3D128&amp;r=G","display_name":"Nik","url":"https:\/\/nikhilc.dev"},{"user_id":257457202,"subscription_id":790270118,"email_address":"michal.dziewonski@automattic.com","date_subscribed":"2024-12-02T14:34:36+00:00","is_email_subscriber":false,"subscription_status":null,"avatar":"https:\/\/2.gravatar.com\/avatar\/5e3c9de90dc73677b682d85fe23af5f477082ab9c1b65b06067ea6e3fa3017d1?s=128&amp;d=https%3A%2F%2F2.gravatar.com%2Favatar%2Fad516503a11cd5ca435acc9bb6523536%3Fs%3D128&amp;r=G","display_name":"Micha\u0142","url":"https:\/\/michaldziewonski.wpcomstaging.com"},{"user_id":179649005,"subscription_id":786241236,"email_address":"sorin.nunca@automattic.com","date_subscribed":"2024-10-19T10:32:44+00:00","is_email_subscriber":false,"subscription_status":null,"avatar":"h`

**Comment by adalpari** [comment]
@jkmassel changed how the `date_subscribed` field is [fa94c4b](https://github.com/Automattic/wordpress-rs/pull/1200/commits/fa94c4b3392062427dc728782313f3ec51b18704). So now, null or malformed entries are treated as a parsing error

File: wp_api/src/wp_com/stats_emails_summary.rs [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+    uniffi::Enum,
+    strum_macros::EnumString,
+    strum_macros::Display,
+)]
+#[serde(rename_all = "snake_case")]
+#[strum(serialize_all = "snake_case")]
+pub enum StatsEmailsSummarySortField {
+    #[default]
+    #[serde(rename = "post_date")]
+    #[strum(serialize = "post_date")]
```
Comment (jkmassel) [inline wp_api/src/wp_com/stats_emails_summary.rs]: You don't need to define `#[serde(rename)]` or `#[strum(serialize)]` again, because you already did `#[serde(rename_all)]` and `#[strum(serialize_all)]`

File: wp_api/src/wp_com/subscribers.rs [hunk 11]
```diff
@@ -198,6 +198,98 @@ pub struct ListSubscribersResponse {
...
+    Hash,
+    Serialize,
+    Deserialize,
+    uniffi::Enum,
+    strum_macros::EnumString,
+    strum_macros::Display,
+)]
+#[serde(rename_all = "snake_case")]
+#[strum(serialize_all = "snake_case")]
+pub enum SubscribersByUserTypeUserType {
```
Comment (jkmassel) [inline wp_api/src/wp_com/subscribers.rs]:
```suggestion
pub enum WPComSubscriberType {
```

WDYT? Not a blocker – if you disagree, feel free too just merge without further review.

**Review by jkmassel — APPROVED** [review]
A couple of minor improvements, but this is ready to go otherwise.

File: wp_api/src/wp_com/stats_emails_summary.rs (hunk 10 above)
Comment (adalpari) [reply wp_api/src/wp_com/stats_emails_summary.rs]: Done!

File: wp_api/src/wp_com/subscribers.rs (hunk 11 above)
Comment (adalpari) [reply wp_api/src/wp_com/subscribers.rs]: SGTM, changed.
